core/tee/tee_rpmb_fs.c

```diff
@@ -1472,54 +1514,364 @@ static TEE_Result tee_rpmb_get_max_block(uint16_t dev_id, uint32_t *max_block)
  */
 
 static TEE_Result get_fat_start_address(uint32_t *addr);
+static TEE_Result rpmb_fs_setup(void);
 
-#if (TRACE_LEVEL >= TRACE_FLOW)
-static void dump_fat(void)
+/**
+ * fat_entry_dir_free: Free the FAT entry dir.
+ * Do not call directly but only within fat_entry_dir_deinit,
+ * or when fat_entry_dir_init fails.
+ */
+static void fat_entry_dir_free(void)
+{
+	if (fat_entry_dir) {
+		if (fat_entry_dir->rpmb_fat_entry_buf)
+			free(fat_entry_dir->rpmb_fat_entry_buf);
+		free(fat_entry_dir);
+		fat_entry_dir = NULL;
+	}
+}
+
+/**
+ * fat_entry_dir_init: Initialize the FAT FS entry buffer/cache
+ * This function must be called before reading FAT FS entries using the
+ * function fat_entry_dir_get_next. This initializes the buffer/cache
+ * with the first FAT FS entries, i.e., 'CFG_RPMB_FS_RD_ENTRIES' when
+ * caching is disabled and
+ * MIN(CFG_RPMB_FS_CACHE_SIZE, CFG_RPMB_FS_RD_ENTRIES) when caching
+ * is enabled. After calling this function, traversing the FAT FS entries
+ * with fat_entry_dir_get_next reads from the buffer/cache and in case
+ * the end of the buffer/cache is reached, more elements are read in from RPMB.
+ * Returns TEE_SUCCESS when buffer/cache initialization succeeds.
+ */
+static TEE_Result fat_entry_dir_init(void)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
-	struct rpmb_fat_entry *fat_entries = NULL;
+	struct rpmb_fat_entry *fe = NULL;
 	uint32_t fat_address;
-	size_t size;
-	int i;
-	bool last_entry_found = false;
+	uint32_t num_elems_read;
+
+	/*
+	 * This holds true when caching is enabled and the cache has already
+	 * been (partly) populated.
+	 */
+	if (fat_entry_dir) {
+		res = TEE_SUCCESS;
+		goto out;
+	}
+
+	res = rpmb_fs_setup();
+	if (res != TEE_SUCCESS)
+		goto out;
 
 	res = get_fat_start_address(&fat_address);
 	if (res != TEE_SUCCESS)
 		goto out;
 
-	size = CFG_RPMB_FS_RD_ENTRIES * sizeof(struct rpmb_fat_entry);
-	fat_entries = malloc(size);
-	if (!fat_entries) {
+	fat_entry_dir = calloc(1, sizeof(struct rpmb_fat_entry_dir));
+	if (!fat_entry_dir) {
 		res = TEE_ERROR_OUT_OF_MEMORY;
 		goto out;
 	}
 
-	while (!last_entry_found) {
-		res = tee_rpmb_read(CFG_RPMB_FS_DEV_ID, fat_address,
-				    (uint8_t *)fat_entries, size, NULL, NULL);
-		if (res != TEE_SUCCESS)
-			goto out;
+	if (CFG_RPMB_FS_CACHE_SIZE == 0)
+		num_elems_read = CFG_RPMB_FS_RD_ENTRIES;
+	/*
+	 * Caching is enabled, so read in up to the maximum cache size, but
+	 * never more than the single read in size.
+	 */
```

Comment:
Please move the comment inside a block. Reversing the condition looks better too.
```C
if (CFG_RPMB_FS_CACHE_SIZE) {
	/*
	 * Caching is enabled, so read in up to the maximum cache size, but
	 * never more than the single read in size.
	 */
	num_elems_read = MIN(CFG_RPMB_FS_CACHE_SIZE,
			     CFG_RPMB_FS_RD_ENTRIES);
} else {
	num_elems_read = CFG_RPMB_FS_RD_ENTRIES;
}
```